File: contests/booking_women_in_tech/b_available.cpp
#include <iostream>
#include <algorithm>

using namespace std;

const int INF = 1000000000;

// [offer type][day][price, min, max stay]
int offers[200][300][3];
// [# nights][day]
int mins[301][300];

int main() {
  int n, m, q;
  cin >> n >> m >> q;

  int x;
  for (int k = 0; k < 3; ++k) {
    for (int i = 0; i < m; ++i) {
      for (int j = 0; j < n; ++j) {
        cin >> x;
        offers[i][j][k] = x;
      }
    }
  }

  for (int i = 0; i < n; ++i) {
    for (int k = 0; k <= n; ++k) {
      int best = INF;

      for (int j = 0; j < m; ++j) {
        if (offers[j][i][0] > 0 && k >= offers[j][i][1] && k <= offers[j][i][2])
          best = min(best, offers[j][i][0]);
      }

      mins[k][i] = best;
    }
  }

  for (int i = 0; i < q; ++i) {
    int c, l;
    cin >> c >> l;

    --c;
    int output = 0;

    for (int i = c; i < c+l; ++i)
      if (mins[l][i] == INF) {
        output = -1;
        break;
      }
      else {
        output += mins[l][i];
      }

    cout << output << endl;
  }

  return 0;
}

File: hackerrank/contests/booking_women_in_tech/b_available.py


INF = 1000000000

n, m, q = [int(x) for x in raw_input().split()]

# (price, min stays, max stays)
offers = [[[0, 0, 0] for _ in range(n)] for _ in range(m)]

for k in range(3):
    for i in range(m):
        line = [int(x) for x in raw_input().split()]

        for j in range(n):
            offers[i][j][k] = line[j]

# (day, n_stays) -> min price
mins = {}

for i in range(n):
    for k in range(n):
        best = INF
        for j in range(m):
            if offers[j][i][0] > 0 and k >= offers[j][i][1] and k <= offers[j][i][2]:
                best = min(best, offers[j][i][0])

        if best != INF:
            mins[(i, k)] = best

for _ in range(q):
    c, l = [int(x) for x in raw_input().split()]
    c -= 1
    output = 0

    for i in range(l):
        r = mins.get((c+i, l))
        if r is None:
            output = -1
            break
        else:
            output += r

    print output
